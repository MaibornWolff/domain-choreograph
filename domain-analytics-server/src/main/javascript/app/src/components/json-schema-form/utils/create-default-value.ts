import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';

const defaultValues: Record<Exclude<JSONSchema4TypeName, 'object'>, any> = {
  string: undefined,
  number: undefined,
  integer: undefined,
  boolean: undefined,
  array: [],
  'null': null,
  any: ''
};

export function createDefaultValue(schema: JSONSchema4): any {
  const type = schema.type || 'any';
  if (type instanceof Array) {
    throw new Error('Array types are not supported');
  } else if (type === 'object') {
    return createDefaultValueForObject(schema);
  } else {
    return defaultValues[type];
  }
}

function createDefaultValueForObject(schema: JSONSchema4): Promise<object> {
  if (schema.type !== 'object') {
    throw new Error('Expect schema to be an object');
  }
  let defaultValue: any = {};
  if (schema.properties) {
    Object.entries(schema.properties).forEach(([key, property]) => {
      defaultValue[key] = createDefaultValue(property)
    });
  }
  return defaultValue;
}

