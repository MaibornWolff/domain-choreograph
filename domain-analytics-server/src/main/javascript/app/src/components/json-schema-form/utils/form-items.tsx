import { styled } from '~utils/styled';

export const FormItems = styled('div')`
    display: flex;
    flex-direction: column;

    & > * {
      margin-bottom: .5rem;
    }
`;
