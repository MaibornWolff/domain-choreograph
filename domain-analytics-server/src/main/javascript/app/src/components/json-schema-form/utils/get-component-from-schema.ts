import { FieldComponent } from 'clean-forms';
import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';
import { DatePickerMsField } from '~components/json-schema-form/form-fields/date-picker-ms-field';
import { NumberTextField, TextField } from '~components/json-schema-form/form-fields/text-field';
import { JsonSchemaFormFieldProps } from '~components/json-schema-form/models/json-schema-form-field-props';

export interface JsonSchemaComponentMap {
  matcher?: { type: JSONSchema4TypeName, format?: string };
  field: FieldComponent<any, JsonSchemaFormFieldProps>;
}

const defaultMapping: JsonSchemaComponentMap[] = [
  { matcher: { type: 'integer', format: 'utc-millisec' }, field: DatePickerMsField },
  { matcher: { type: 'integer' }, field: NumberTextField },
  { matcher: { type: 'number' }, field: NumberTextField },
  { field: TextField }
];

export function getComponentFromSchema(
  schema: JSONSchema4,
  mapping: JsonSchemaComponentMap[] = defaultMapping
): FieldComponent<any, JsonSchemaFormFieldProps> {
  for (const map of mapping) {
    if (map.matcher == null) {
      return map.field;
    }
    const typeIsMatching = map.matcher.type === schema.type;
    const formatIsMatching = map.matcher.format == null || map.matcher.format === schema.format;
    if (typeIsMatching && formatIsMatching) {
      return map.field;
    }
  }
  throw new Error('Couldn\'t match the schema');
}
