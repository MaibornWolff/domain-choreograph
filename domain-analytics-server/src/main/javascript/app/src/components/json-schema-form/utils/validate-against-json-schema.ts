import { ValidationFunction } from 'clean-forms';
import { JSONSchema4 } from 'json-schema';
import { Validator } from 'jsonschema';


export function validateAgainstJsonSchema(schema: JSONSchema4): ValidationFunction {
  const validator = new Validator();
  return ({ value }) => {
    const result = validator.validate(value, schema as any);
    const errors = result.errors.map(error => {
      let message = error.message;
      let property = error.property
        .replace(/^instance\.?/, '');
      let schema = error.schema as JSONSchema4;
      if (error.name === 'required') {
        message = `is required`;
        schema = schema.properties![error.argument];
        property = joinProperties(property, error.argument);
      }

      const title = schema.title || '';
      message = `${title} ${message}`;

      return [ property, message ] as [string, string];
    });
    return errors;
  }
}

function joinProperties(path1: string, path2: string): string {
  if (path1 === '') {
    return path2;
  }
  if (path2 === '') {
    return path1;
  }
  return `${path1}.${path2}`;
}
