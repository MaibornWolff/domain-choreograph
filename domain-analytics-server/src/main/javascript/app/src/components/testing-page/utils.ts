export function escapeJavaClass(javaClass: string) {
  return javaClass.replace(/\./g, '-');
}

export function reverseJavaClassEscape(javaClass: string) {
  return javaClass.replace(/-/g, '.');
}
