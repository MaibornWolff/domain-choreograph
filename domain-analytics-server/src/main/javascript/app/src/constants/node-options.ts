export const NODE_OPTIONS = {
  minHeightInPx: 50,
  maxWidthInPx: 250,
  sideMarginInPx: 16,
};

export type NodeOptions = typeof NODE_OPTIONS;
