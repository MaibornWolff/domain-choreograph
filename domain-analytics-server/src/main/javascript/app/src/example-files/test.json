{
  "nodes": {
    "A": {
      "id": "A",
      "name": "A",
      "scope": "root"
    }
  },
  "dependencies": [],
  "scopes": {
    "root": {
      "id": "root",
      "executionContext": "Application",
      "nodes": [
        "A"
      ]
    }
  },
  "executionContexts": {
    "Application": {
      "id": "Application",
      "scopes": [
        "root"
      ]
    }
  }
}
