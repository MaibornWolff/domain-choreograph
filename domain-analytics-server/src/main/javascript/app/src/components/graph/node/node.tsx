import { transparentize } from 'polished';
import * as React from 'react';
import { pure } from 'recompose';
import { SelectedOptions } from '~components/graph/models/selected-options';
import { NodeLink } from '~components/graph/node-link/node-link';
import { NodeOptions } from '~constants/node-options';
import { Theme } from '~styles/theme';
import { ChoreoExecutionContext, ChoreoNode } from '~types/choreo-graph';
import { selectPreviewTextFromNode } from '~utils/node.utils';
import { styled } from '~utils/styled';

export interface NodeModel extends ChoreoNode {
  x: number;
  y: number;
  height: number;
  width: number;
  outsideScope: boolean;
  executionContext?: ChoreoExecutionContext;
}

export interface NodeProps {
  node: NodeModel;
  scale: number;
  options: NodeOptions;
  isSelected: boolean;
  selectedOptions?: SelectedOptions;
  toggleNode?: (nodeId: string) => void;
}

function getBorderColor(args: { theme: Theme } & NodeProps) {
  const { theme, node, isSelected, selectedOptions = {} } = args;
  if (node.hasException) {
    return theme.colors.danger;
  }
  if (isSelected) {
    return selectedOptions.color || theme.colors.selectedNode;
  }
  return theme.colors.border.normal;
}

const Wrapper = styled.div<NodeProps>`
  position: absolute;
  background: ${({ theme }) => theme.colors.background[3]};
  border: ${(args) => `
    ${args.node.outsideScope ? 'dashed' : 'solid'}
    ${args.theme.dimensions.border.thick}
    ${getBorderColor(args)}
  `};
  ${(args) => args.isSelected
  ? `box-shadow: 0 0 5px 1px ${transparentize(.5, getBorderColor(args))}`
  : `box-shadow: 0 0 5px 1px ${transparentize(.15, getBorderColor(args))}`
  };
  border-radius: ${({ theme }) => theme.dimensions.borderRadius.big};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 0 ${({ scale, options }) => options.sideMarginInPx * scale}px;
  width: ${({ node, scale }) => node.width * scale}px;
  height: ${({ node, scale }) => node.height * scale}px;
  font-size: calc(1rem * ${({ scale }) => scale});
  box-sizing: border-box;
  left: ${({ node, scale }) => (node.x - node.width / 2) * scale}px;
  top: ${({ node, scale }) => (node.y - node.height / 2) * scale}px;
  cursor: pointer;
  transition: ${({theme}) => theme.animations.all};
  z-index: 1;
`;

const ValueWrapper = styled.pre<NodeProps>`
  margin: ${({ scale }) => scale * .3}rem;
  text-align: center;
  font-size: ${({ scale }) => scale * .8}rem;
  color: ${({ node, theme }) => node.hasException ? theme.colors.danger : theme.colors.text.normal.secondary};
  white-space: pre-wrap;
  width: 100%;
  overflow: hidden;
`;

const NameWrapper = styled.div`
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
`;

const _Node = (props: NodeProps) => {
  props = {
    ...props,
    scale: props.scale || 1
  };
  const { node, toggleNode } = props;
  const previewMessage = selectPreviewTextFromNode(node);
  return <Wrapper {...props} onClick={() => toggleNode && toggleNode(node.id)}>
    {node.executionContext ? <NodeLink executionContextId={node.executionContext.id}/> : null}
    <NameWrapper>{node.name}</NameWrapper>
    {previewMessage == null ? null : <ValueWrapper {...props}>{previewMessage}</ValueWrapper>}
  </Wrapper>;
};

export const Node = pure(_Node);
