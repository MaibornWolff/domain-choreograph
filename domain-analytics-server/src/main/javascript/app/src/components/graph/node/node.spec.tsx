import * as React from 'react';
import { NODE_OPTIONS } from '~constants/node-options';
import { renderWithProviders } from '~utils/test.utils';
import { Node, NodeModel } from './node';

describe('Node', () => {
  it('should render', () => {
    const node: NodeModel = {
      id: 'test',
      name: 'test',
      scope: 'test',
      x: 100,
      y: 200,
      height: 10,
      width: 20,
      value: { hello: 'world' },
      outsideScope: false
    };
    expect(renderWithProviders(<Node node={node} scale={.5} options={NODE_OPTIONS} isSelected={false}/>)).toMatchSnapshot();
  });

  it('should render if selected', () => {
    const node: NodeModel = {
      id: 'test',
      name: 'test',
      scope: 'test',
      x: 100,
      y: 200,
      height: 10,
      width: 20,
      value: { hello: 'world' },
      outsideScope: false
    };
    expect(renderWithProviders(<Node node={node} scale={.5} options={NODE_OPTIONS} isSelected={true}/>)).toMatchSnapshot();
  });

  it('should render with preview', () => {
    const node: NodeModel = {
      id: 'test',
      name: 'test',
      scope: 'test',
      preview: 'value',
      x: 100,
      y: 200,
      height: 10,
      width: 20,
      value: { hello: 'world' },
      outsideScope: false
    };
    expect(renderWithProviders(<Node node={node} scale={.5} options={NODE_OPTIONS} isSelected={false}/>)).toMatchSnapshot();
  });
});
