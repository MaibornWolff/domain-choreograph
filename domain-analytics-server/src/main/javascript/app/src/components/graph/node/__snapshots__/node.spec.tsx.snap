// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node should render 1`] = `
.c0 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8px;
  width: 10px;
  height: 5px;
  font-size: calc(1rem * 0.5);
  box-sizing: border-box;
  left: 45px;
  top: 97.5px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c1 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

<div
  class="c0"
  scale="0.5"
>
  <div
    class="c1"
  >
    test
  </div>
</div>
`;

exports[`Node should render if selected 1`] = `
.c0 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px #4d66b3;
  box-shadow: 0 0 5px 1px rgba(77,102,179,0.5);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8px;
  width: 10px;
  height: 5px;
  font-size: calc(1rem * 0.5);
  box-sizing: border-box;
  left: 45px;
  top: 97.5px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c1 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

<div
  class="c0"
  scale="0.5"
>
  <div
    class="c1"
  >
    test
  </div>
</div>
`;

exports[`Node should render with preview 1`] = `
.c0 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8px;
  width: 10px;
  height: 5px;
  font-size: calc(1rem * 0.5);
  box-sizing: border-box;
  left: 45px;
  top: 97.5px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c2 {
  margin: 0.15rem;
  text-align: center;
  font-size: 0.4rem;
  color: rgba(0,0,0,.54);
  white-space: pre-wrap;
  width: 100%;
  overflow: hidden;
}

.c1 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

<div
  class="c0"
  scale="0.5"
>
  <div
    class="c1"
  >
    test
  </div>
  <pre
    class="c2"
    scale="0.5"
  >
    value
  </pre>
</div>
`;
